### PR TITLE
[EE-3024] remove dead code

### PR DESCRIPTION
**Pull request by thomaschaplin**
# Dead Code Analysis: Removed arg.selectionSet Branch

## Overview

The removed code was a conditional branch in the `getArguments` function that checked for `arg.selectionSet`:

```typescript
// REMOVED DEAD CODE (was line 109-110)
} else if (arg.selectionSet) {
    argsObj[arg.name.value] = getSelections(arg.selectionSet.selections)
```

This code is **provably dead** and safe to remove based on GraphQL specification, AST structure, and empirical testing.

### Argument Node Structure
In the GraphQL AST, argument nodes have a strictly defined structure:

```typescript
interface ArgumentNode {
    kind: "Argument"
    name: NameNode
    value: ValueNode
    loc?: Location
}
```

**Key Point**: Argument nodes do NOT have a `selectionSet` property.

### Empirical Testing
I tested this with actual GraphQL queries using the `graphql` library's `parse` function:

```javascript
const { parse } = require('graphql');

const query = `
query {
  user(id: "123", filter: { name: "John", active: true }) {
    name
    email
  }
}`;

const ast = parse(query);
const firstField = ast.definitions[0].selectionSet.selections[0];
firstField.arguments.forEach(arg => {
    console.log(`Argument "${arg.name.value}" properties:`, Object.keys(arg));
    console.log(`Has selectionSet: ${!!arg.selectionSet}`);
});
```

**Output:**
```
Argument "id" properties: ['kind', 'name', 'value', 'loc']
Has selectionSet: false

Argument "filter" properties: ['kind', 'name', 'value', 'loc']
Has selectionSet: false
```

### Invalid Syntax Examples
Arguments cannot contain selection sets according to GraphQL syntax rules. Attempts to create such syntax result in parse errors:

```graphql
# INVALID - This causes a syntax error
query {
  user(profile { name email }) {  # ❌ Syntax Error: Expected ":", found "{"
    id
  }
}

# INVALID - Arguments don't have selection sets
query {
  user(id: "123" { subfield }) {  # ❌ Parse error
    name
  }
}
```

### Valid GraphQL Argument Patterns
Valid arguments can only be:
- Scalars: `field(count: 10)`
- Objects: `field(filter: { name: "John" })`
- Lists: `field(tags: ["a", "b"])`
- Enums: `field(status: ACTIVE)`
- Variables: `field(id: $userId)`

**None of these patterns create an `arg.selectionSet` property.**

### Selection Sets vs Arguments
- **Selection Sets** define *what fields to query* (the `{ name email }` part)
- **Arguments** define *parameters for fields* (the `(id: "123")` part)

These are fundamentally different concepts in GraphQL:

```graphql
query {
  user(id: "123") {  # ← Arguments: parameters
    name             # ← Selection Set: fields to query
    email
  }
}
```

### AST Node Hierarchy
```
Query
└── Field "user"
    ├── Arguments: [{ name: "id", value: "123" }]  # ← No selectionSet here
    └── SelectionSet                               # ← selectionSet is here
        ├── Field "name"
        └── Field "email"
```

### Reachability Analysis
For `arg.selectionSet` to be truthy, one of these would need to be true:

1. **GraphQL parser adds selectionSet property to arguments** → ❌ Never happens
2. **Malformed AST manipulation** → ❌ Would break other parts first
3. **Custom AST construction** → ❌ Library only accepts parsed GraphQL strings

### Control Flow Analysis
The `getArguments` function processes each argument through these conditions:

```typescript
args.forEach((arg) => {
    if (arg.value.kind === "ObjectValue") {        // ✅ Handles objects
        // ...
    } else if (arg.value.kind === "Variable") {    // ✅ Handles variables
        // ...
    } else if (arg.selectionSet) {                 // ❌ NEVER REACHED
        // DEAD CODE
    } else if (arg.value.kind === "EnumValue") {   // ✅ Handles enums
        // ...
    } else if (arg.value.kind === "IntValue") {    // ✅ Handles integers
        // ...
    } else if (arg.value.kind === "ListValue") {   // ✅ Handles arrays
        // ...
    } else {                                       // ✅ Handles strings/other
        // ...
    }
})
```

Since `arg.selectionSet` is always `undefined` for valid GraphQL arguments, this branch is unreachable.

### Before Removal
- **Coverage**: 98.96% statements, 98% branches
- **Missing**: Line 110 and the `arg.selectionSet` branch

### After Removal
- **Coverage**: 100% statements, 100% branches
- **Tests**: All 55 tests pass
- **Functionality**: Completely unchanged

This proves the code was never executed in any test scenario.

## Evidence 6: Edge Case Testing

I tested various edge cases that might theoretically trigger this code:

### Fragments
```graphql
fragment UserFragment on User { name email }
query { user(id: "123") { ...UserFragment } }
```
**Result**: Arguments still have no `selectionSet` property.

### Nested Objects
```graphql
query { user(filter: { profile: { name: "John" } }) { id } }
```
**Result**: Nested objects are handled by `ObjectValue` branch, not `selectionSet`.

### Complex Scenarios
```graphql
query { 
  posts(
    filter: { 
      author: { name: "John", posts: { status: PUBLISHED } }
    }
  ) { 
    title 
  } 
}
```
**Result**: All arguments processed correctly without touching the dead code path.

### Why Removal is Safe

1. **No Functionality Loss**: The code was never executed, so removing it changes nothing
2. **No Breaking Changes**: All existing behavior is preserved
3. **Improved Maintainability**: Eliminates confusion about unreachable code
4. **Performance**: Tiny improvement by removing unnecessary condition check

### Backwards Compatibility
- **API**: No changes to public interface
- **Behavior**: Identical output for all inputs
- **Dependencies**: No changes to external dependencies

## Test Coverage Impact

The removal of dead code resulted in **perfect 100% test coverage**:

| Metric | Before | After | Change |
|--------|--------|-------|---------|
| Statements | 98.96% (96/97) | 100% (95/95) | +1.04% |
| Branches | 98% (49/50) | 100% (48/48) | +2% |
| Functions | 100% (19/19) | 100% (19/19) | No change |
| Lines | 98.91% (91/92) | 100% (90/90) | +1.09% |

## Conclusion

The removed code is definitively dead because:

1. **Structurally impossible**: GraphQL AST arguments cannot have `selectionSet` properties
2. **Specification compliance**: GraphQL syntax rules prevent such constructs  
3. **Empirically verified**: Extensive testing confirms the condition is never true
4. **Logically sound**: The concepts of arguments and selection sets are orthogonal

Removing this code improves code quality by eliminating dead branches while maintaining 100% functional compatibility. The change is completely safe and has been validated through comprehensive testing.

## Additional Context

This analysis was conducted as part of improving test coverage from 98.96% to 100%. The dead code was identified through coverage analysis and confirmed through multiple verification methods including:

- Static analysis of GraphQL AST structure
- Dynamic testing with various query patterns
- Specification review
- Empirical validation with the `graphql` library

The removal contributes to better code maintainability and eliminates a potential source of confusion for future developers.

---

*Analysis conducted using graphql@16.x parsing behavior and validated against the GraphQL specification.*

### CODE REVIEW

**Comment by thomaschaplin** [comment]
@rickschubert please could you review?

File: src/index.ts [hunk 1]
```diff
@@ -106,8 +106,6 @@ const getArguments = (args) => {
         } else if (arg.value.kind === "Variable") {
             argsObj[arg.name.value] =
                 `${arg.value.name.value}${isVariableDropinConst}`
-        } else if (arg.selectionSet) {
-            argsObj[arg.name.value] = getSelections(arg.selectionSet.selections)
```
Comment (thomaschaplin) [inline src/index.ts:110]: Additionally...

getSelections is used in two places:

  1. Line 134: Recursively called within itself to handle nested selection sets in
   fields:
  if (selection.selectionSet) {
      selObj[selectionName] = getSelections(selection.selectionSet.selections)
  }
  2. Line 222: Called from the main graphQlQueryToJson function to process the
  root selection set:
  const selections = getSelections(firstDefinition.selectionSet.selections)

  The function processes field selections (like user { name email }), not argument
   selections. Fields can have nested selection sets, but arguments cannot - which
   is why the arg.selectionSet branch was dead code.

**Review by rickschubert — APPROVED** [review]
Amazing work, well done -- approved!

My initial gut reaction was "no, let's leave it there, better safe than sorry", but you brought numerous convincing arguments to the table. On top of that, our test suite is very extensive (even made better now by these new tests you added). Therefore, if this line was truly needed, we would have hit it eventually; but we don't. So it surely seems not needed.

Well done!

**Review by jdeastwood — APPROVED** [review]
